core/src/main/java/hudson/model/Computer.java

```diff
@@ -1093,6 +1095,7 @@ public final long getDemandStartMilliseconds() {
      */
     protected void removeExecutor(final Executor e) {
         final Runnable task = () -> {
```

Comment:
Hmm, now I see that this whole `task` is also inside the `Queue` lock too. Would be safest to move notification outside there, but then you have to make sure notification happens after the `remove` calls, so it is looking easier to going back to asynch notification. Sorry.